ThePhasesOfCraftship/1_code_first/assignment/begin/backend/src/index.ts

```diff
@@ -0,0 +1,186 @@
+            username,
+            firstName,
+            lastName,
+            password: basicPasswordGenerator(10)
+        }
+    })
+    return user
+}
+
+const removePasswordForResponse = (user:  User): UserResponse => {
+    const { password, ...userResponse } = user
+    return userResponse
+}
+
+const getUserByEmail = async (email: string) => {    
+    const user = await prisma.user.findUnique({
+        where: {
+            email
+        }
+    })
+    return user
+
+}
+
+
+const updateUser = async (userId: number, { email, firstName, lastName, username }: Omit<User, 'password'>) => {
+    const user = await prisma.user.update({
+        where: {
+            id: userId
+        },
+        data: {
+            email,
+            firstName,
+            lastName,
+            username
+        }
+    })
+    return user
+}
+
+const app = express()
+const port = 8090
+
+app.use(express.json())
+
+app.listen(port, () => {
+    console.log(`Server is running on port ${port}`)
+})
+
+app.get('/', (req: Request, res: Response) => {
+    res.send('Landing Page')
+})
+
+// Create a new user - spaghetti code example
+app.post('/users/new', async (req: Request, res: Response, next: NextFunction) => {
+    const { email, firstName, lastName, username } = req.body
+    try {
+        const user = await createUser({ email, firstName, lastName, username })
+        const response = responseObjectFactory(undefined, removePasswordForResponse(user), true)
+        res.status(201).send(response)
+    } catch (error) {
+        console.log('error handler')
+        next(error)
+    }
+});
+        
+        // Edit a user
+app.post('/users/edit/:userId', async (req: Request, res: Response, next: NextFunction) => {
+    // ...
+    const userId = parseInt(req.params.userId)
+    const { email, firstName, lastName, username } = req.body    
+    try {
+        const user = await updateUser(userId, { email, firstName, lastName, username })
+        if (!user) {
+            const userNotFoundResponse = responseObjectFactory(ErrorName.UserNotFound, undefined, false)
+            return res.status(404).json(userNotFoundResponse)
+        }
+        const response = responseObjectFactory(undefined, removePasswordForResponse(user), true)
+        res.status(200).send(response)
+    } catch (error) {
+        next(error)
+    }
+    
+    
+});
+
+// Get a user by email
+app.get('/users', async (req: Request, res: Response, next: NextFunction) => {
+    const email = req.query.email as string
+    try {
+        const user = await getUserByEmail(email)        
+        if (!user) {
+            const userNotFoundResponse = responseObjectFactory(ErrorName.UserNotFound, undefined, false)
+            return res.status(404).json(userNotFoundResponse)
+        }
+        const response = responseObjectFactory(undefined, removePasswordForResponse(user), true) 
+        res.status(200).send(response)
+    } catch (error) {
+        next(error)
```

Comment:
Due to my lack of experience with express and BE in general, I also had a hard time figuring out why you keep calling `next` in `catch`. If I understand it correctly you did this in order to allow centralized error handling with middleware.
It looks like an interesting approach to me, not sure if I love it though. I'd be curious what others think, because if this is a best practice I'd be happy to dig into it deeper :)